runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperPlugin.java

```diff
@@ -0,0 +1,394 @@
+	}
+
+	@Subscribe
+	public void onPlayerMenuOptionClicked(PlayerMenuOptionClicked event)
+	{
+		if (!event.getMenuOption().equals(HOP_TO))
+		{
+			return;
+		}
+
+		ChatPlayer player = getChatPlayerFromName(event.getMenuTarget());
+
+		if (player != null)
+		{
+			hop(player.getWorld());
+		}
+	}
+
+	private void hop(boolean previous)
+	{
+		if (worldResult == null || client.getGameState() != GameState.LOGGED_IN)
+		{
+			return;
+		}
+
+		int totalLevel = client.getTotalLevel();
+
+		World currentWorld = worldResult.findWorld(client.getWorld());
+
+		// Make it so you always hop out of PVP, high risk and bounty worlds
+		// Also don't regard skill total worlds as a type that must be hopped between
+		EnumSet<WorldType> currentWorldTypes = currentWorld.getTypes().clone();
+		currentWorldTypes.remove(WorldType.PVP);
+		currentWorldTypes.remove(WorldType.PVP_HIGH_RISK);
+		currentWorldTypes.remove(WorldType.BOUNTY);
+		currentWorldTypes.remove(WorldType.SKILL_TOTAL);
+
+		List<World> worlds = worldResult.getWorlds();
+
+		int worldIdx = worlds.indexOf(currentWorld);
+
+		World world;
+		do
+		{
+			if (previous)
+			{
+				worldIdx--;
+			}
+			else
+			{
+				worldIdx++;
+			}
+
+			if (worldIdx < 0)
+			{
+				worldIdx = worlds.size() - 1;
+			}
+			else if (worldIdx >= worlds.size())
+			{
+				worldIdx = 0;
+			}
+
+			world = worlds.get(worldIdx);
+
+			EnumSet<WorldType> types = world.getTypes().clone();
+
+			if (types.contains(WorldType.SKILL_TOTAL))
+			{
+				try
+				{
+					int totalRequirement = Integer.parseInt(world.getActivity().substring(0, world.getActivity().indexOf(" ")));
+
+					if (totalLevel >= totalRequirement)
+					{
+						types.remove(WorldType.SKILL_TOTAL);
+					}
+				}
+				catch (NumberFormatException ex)
+				{
+					log.error(null, ex);
+				}
+			}
+
+			if (currentWorldTypes.equals(types))
+			{
+				break;
+			}
+		}
+		while (world != currentWorld);
+
+		if (world == currentWorld)
+		{
+			String chatMessage = new ChatMessageBuilder()
+					.append(ChatColorType.NORMAL)
+					.append("Couldn't find a world to quick-hop to.")
+					.build();
+
+			chatMessageManager.queue(QueuedMessage.builder()
+					.type(ChatMessageType.GAME)
+					.runeLiteFormattedMessage(chatMessage)
+					.build());
+		}
+		else
+		{
+			hop(world.getId());
+		}
+	}
+
+	private void hop(int world)
+	{
+		// Don't try to hop if the world doesn't exist
+		if (worldResult.findWorld(world) == null)
+		{
+			return;
+		}
+
+		worldToHopTo = world;
+
+		String chatMessage = new ChatMessageBuilder()
+				.append(ChatColorType.NORMAL)
+				.append("Quick-hopping to World ")
+				.append(ChatColorType.HIGHLIGHT)
+				.append(worldToHopTo + "")
+				.append(ChatColorType.NORMAL)
+				.append("..")
+				.build();
+
+		chatMessageManager
+				.queue(QueuedMessage.builder()
+				.type(ChatMessageType.GAME)
+				.runeLiteFormattedMessage(chatMessage)
+				.build());
+	}
+
+	@Subscribe
+	public void onGameTick(GameTick event)
```

Comment:
Why is this onGameTick? Can't you just use ClientThread.invokeLater instead where you setting the world to hop to and hop directly there?

Reply:
Hopping in ClientThread.invokeLater doesn't work without waiting anything since the world hopper was opened. I think the source of the issue is that if both openWorldHopper and hopToWorld are parsed in the same tick on the server, it doesn't let you hop.

Reply:
You can return true from ClientThread.invokeLater and have the method rescheduled for next tick. So you can do same logic as here, but instead of having worldToHopto reset to -1, you just return false from .invokeLater (and true when opening worldHopper)

Reply:
Changed it to do that now, seems to work good.